src/protocol/tcp/src/main/java/org/apache/jmeter/protocol/tcp/proxy/TCPSamplerManager.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.jmeter.testelement.TestElement;
+import org.apache.jorphan.util.JOrphanUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TCPSamplerManager extends Thread {
+    private static final Logger log = LoggerFactory.getLogger(TCPSamplerManager.class);
+
+    private final JMeterTreeNode targetNode;
+    private final TestElement tcpSamplerElement;
+    private boolean workStatus = false;
+
+    private final Queue<TCPSampler> tcpSamplerInfoQueue = new ArrayDeque<>();
+
+    public TCPSamplerManager(TestElement tcpSamplerElement, JMeterTreeNode targetNode) {
+        this.tcpSamplerElement = tcpSamplerElement;
+        this.targetNode = targetNode;
+    }
+
+    /**
+     * create a new TCPSampler and put into TCPSampler queue.
+     * manager will store all TCPSampler to target JmeterTreeNode.
+     *
+     * @param data tcp payload
+     * @return target server response payload
+     */
+    public byte[] newTCPSampler(String data) {
+        log.debug("new sampler , payload " + data);
+        TCPSampler tcpSampler = new TCPSampler();
+        TestElement element = (TestElement) tcpSamplerElement.clone();
+        element.removeProperty(TCPSampler.PROXY_SERVER_PORT);
+        if (useHexPayload(element)) {
+            data = JOrphanUtils.baToHexString(data.getBytes());
+        }
+        element.setProperty(TCPSampler.REQUEST, data);
+        element.setProperty(TestElement.GUI_CLASS, TCPSamplerGui.class.getName());
+        element.setName("TCP Sampler");
+        tcpSampler.addTestElement(element);
+        tcpSamplerInfoQueue.add(tcpSampler);
+        SampleResult tcpSampleResult = tcpSampler.sample();
+        if (useHexPayload(element)) {
+            return BinaryTCPClientImpl.hexStringToByteArray(new String(tcpSampleResult.getResponseData()));
+        }
+        return tcpSampleResult.getResponseData();
+    }
+
+    private boolean useHexPayload(TestElement element) {
+        String tcpSamplerClassPath = element.getPropertyAsString(TCPSampler.CLASSNAME);
+        TCPProxyDef samplerClass = TCPProxyDef.findByClassPath(tcpSamplerClassPath);
+        if (samplerClass != null) {
+            switch (samplerClass) {
+                case BinaryTCPClientImpl_class:
+                case LengthPrefixedBinaryTCPClientImpl_class:
+                    return true;
+                case TCPClientImpl_class:
+                default:
+                    return false;
+            }
+        }
+        return false;
+    }
+
+    public synchronized void managerStart() {
+        log.debug("start");
+        workStatus = true;
+        start();
+    }
+
+    /**
+     * manager will stop after samplerQueue is empty
+     */
+    public void managerStop() {
+        workStatus = false;
+    }
+
+    @Override
+    public void run() {
+        while (workStatus) {
+            try {
+                Thread.sleep(500);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```

Comment:
prefer logging

Reply:
I have added the " Exception Log" in new commit.